File: sampo/native/time.cpp
#include "time.h"

#define TIME_INF 2000000000

class Time {
private:
    int value;
public:
    explicit Time(int value) {
        if (value > TIME_INF) {
            value = TIME_INF;
        } else if (value < 0) {
            value = 0;
        }
        this->value = value;
    }

    static Time inf() {
        static Time inf(TIME_INF); // caching the inf instance
        return inf;
    }

    bool is_inf() const {
        return this->value == TIME_INF;
    }

    // Copy-paste body between functions with different arg type
    // is here to avoid functional call overhead. Time instances
    // should be as fast as possible.

    Time operator+(Time& other) const {
        return Time(this->value + other.value);
    }

    Time operator+(int other) const {
        return Time(this->value + other);
    }

    Time operator-(Time& other) const {
        return Time(this->value - other.value);
    }

    Time operator-(int other) const {
        return Time(this->value - other);
    }

    Time operator*(Time& other) const {
        return Time(this->value * other.value);
    }

    Time operator*(int other) const {
        return Time(this->value * other);
    }

    Time operator/(Time& other) const {
        return Time(this->value / other.value);
    }

    Time operator/(int other) const {
        return Time(this->value / other);
    }

    bool operator<(Time& other) const {
        return this->value < other.value;
    }

    bool operator<(int other) const {
        return this->value < other;
    }

    bool operator>(Time& other) const {
        return this->value > other.value;
    }

    bool operator>(int other) const {
        return this->value > other;
    }

    bool operator<=(Time& other) const {
        return this->value <= other.value;
    }

    bool operator<=(int other) const {
        return this->value <= other;
    }

    bool operator>=(Time& other) const {
        return this->value >= other.value;
    }

    bool operator>=(int other) const {
        return this->value >= other;
    }

    bool operator==(Time& other) const {
        return this->value == other.value;
    }

    bool operator==(int other) const {
        return this->value == other;
    }
};

File: sampo/schemas/time.py
from sampo.schemas.serializable import AutoJSONSerializable

TIME_INF = 2_000_000_000


# TODO Consider converting to StrSerializable
class Time(AutoJSONSerializable['Time']):
    """
    Class for describing all basic operations for working with time in framework

    :param value: initial time value
    """
    value: int

    def __init__(self, value: int = 0):
        """
        :param value: value of time
        """
        self.value = 0
        self.set_time(value)

    @staticmethod
    def inf():
        """
        Return time, that is obviously longer
        """
        return Time(TIME_INF)

    # Copy-paste body between left- and right-associative functions
    # is here to avoid functional call overhead
    def __add__(self, other: 'Time' or int) -> 'Time':
        return Time(self.value + (other.value if isinstance(other, Time) else other))

    def __radd__(self, other: 'Time' or int) -> 'Time':
        return Time(self.value + (other.value if isinstance(other, Time) else other))

    def __sub__(self, other: 'Time' or int) -> 'Time':
        return Time(self.value - (other.value if isinstance(other, Time) else other))

    def __rsub__(self, other: 'Time' or int) -> 'Time':
        return Time((other.value if isinstance(other, Time) else other) - self.value)

    def __mul__(self, other: 'Time' or int) -> 'Time':
        return Time(self.value * (other.value if isinstance(other, Time) else other))

    def __rmul__(self, other: 'Time' or int) -> 'Time':
        return Time(self.value * (other.value if isinstance(other, Time) else other))

    def __floordiv__(self, other: 'Time' or int) -> 'Time':
        return Time(self.value // (other.value if isinstance(other, Time) else other))

    def __truediv__(self, other: 'Time' or int) -> float:
        return self.value / (other.value if isinstance(other, Time) else other)

    def __lt__(self, other):
        return self.value < other

    def __le__(self, other):
        return self.value <= other

    def __gt__(self, other):
        return self.value > other

    def __ge__(self, other):
        return self.value >= other

    def __eq__(self, other):
        return self.value == other

    def __bool__(self):
        return self.value != 0

    def __int__(self) -> int:
        return self.value

    def __str__(self) -> str:
        return str(self.value)

    def __repr__(self) -> str:
        return str(self.value)

    def __unicode__(self) -> str:
        return str(self.value)

    def __hash__(self) -> int:
        return hash(self.value)

    def set_time(self, value: int):
        value = int(value)
        if value > TIME_INF:
            value = TIME_INF
        elif value < -TIME_INF:
            value = -TIME_INF
        self.value = value

    def is_inf(self) -> bool:
        return abs(self.value) == TIME_INF
